### PR TITLE
Renaming Agent's methods

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

* The following methods in the `Agent` class were renamed:
   * `InitializeAgent()` was renamed to `Initialize()`
   * `AgentAction()` was renamed to `OnActionReceived()`
   * `AgentReset()` was renamed to `OnEpsiodeBegin()`
   * `Done()` was renamed to `EndEpisode()`
   * `GiveModel()` was renamed to `SetModel()`

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

[Renaming sheet](https://docs.google.com/spreadsheets/d/1ohos7YHyLrpD2p-RmHsy0OH_wFpnZIVaFyZkdm0UJYc/edit#gid=685061562)

### Types of change(s)

- ~[ ] Bug fix~
- ~[ ] New feature~
- [x] Code refactor
- [x] Breaking change
- [x] Documentation update
- ~[ ] Other (please describe)~

### Checklist
- ~[ ] Added tests that prove my fix is effective or that my feature works~
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [x] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 1]
```diff
@@ -267,7 +267,10 @@ public void LazyInitialize()
             Academy.Instance.AgentForceReset += _AgentReset;
             m_Brain = m_PolicyFactory.GeneratePolicy(Heuristic);
             ResetData();
+#pragma warning disable 0618
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Agent.cs]: This is a warning about calling an obsolete method, right?

Would it be cleaner to only call the new version here, and make the default implementation of `Initialize` call `InitializeAgent` (wrapped in the pragmas)? That way there's no chance that someone accidentally implements both methods and we call them both

Comment (chriselion) [reply com.unity.ml-agents/Runtime/Agent.cs]: (same comments for AgentReset and AgentAction too)

File: docs/Getting-Started-with-Balance-Ball.md [hunk 2]
```diff
@@ -60,31 +60,31 @@ behavior:
...
   training generalizes to more than a specific starting position and agent cube
   attitude.
-* agent.CollectObservations(VectorSensor sensor) — Called every simulation step. Responsible for
+* `Agent.CollectObservations(VectorSensor sensor)` — Called every simulation step. Responsible for
   collecting the Agent's observations of the environment. Since the Behavior
   Parameters of the Agent are set with vector observation
   space with a state size of 8, the `CollectObservations(VectorSensor sensor)` must call
   `VectorSensor.AddObservation()` such that vector size adds up to 8.
-* agent.AgentAction() — Called every simulation step. Receives the action chosen
+* `Agent.OnActionReceived()` — Called every simulation step. Receives the action chosen
```
Comment (chriselion) [inline docs/Getting-Started-with-Balance-Ball.md]: Not directly related to this PR, but "Called every simulation step" is not correct.

File: docs/Getting-Started-with-Balance-Ball.md [hunk 3]
```diff
@@ -60,31 +60,31 @@ behavior:
...
   by the Agent. The vector action spaces result in a
-  small change in the agent cube's rotation at each step. The `AgentAction()` function
+  small change in the agent cube's rotation at each step. The `OnActionReceived()` function
   assigns a reward to the Agent; in this example, an Agent receives a small
   positive reward for each step it keeps the ball on the agent cube's head and a larger,
-  negative reward for dropping the ball. An Agent is also marked as done when it
+  negative reward for dropping the ball. An Agent episode is also ended when it
```
Comment (chriselion) [inline docs/Getting-Started-with-Balance-Ball.md]:
```suggestion
  negative reward for dropping the ball. An Agent's episode is also ended when it
```

File: docs/Getting-Started-with-Balance-Ball.md [hunk 4]
```diff
@@ -60,31 +60,31 @@ behavior:
...
 * **Max Step** — Defines how many simulation steps can occur before the Agent
-  decides it is done. In 3D Balance Ball, an Agent restarts after 5000 steps.
+  episode ends. In 3D Balance Ball, an Agent restarts after 5000 steps.
 
 Perhaps the more interesting aspect of an agents is the Agent subclass
 implementation. When you create an Agent, you must extend the base Agent class.
 The Ball3DAgent subclass defines the following methods:
 
-* agent.AgentReset() — Called when the Agent resets, including at the beginning
+* `Agent.OnEpisodeBegin()` — Called when the Agent resets, including at the beginning
```
Comment (chriselion) [inline docs/Getting-Started-with-Balance-Ball.md]: "... including at the beginning of simulation"?

File: docs/Learning-Environment-Create-New.md [hunk 5]
```diff
@@ -128,7 +128,7 @@ to the cube using reinforcement learning.
 
 ### Initialization and Resetting the Agent
 
-When the Agent reaches its target, it marks itself done and its Agent reset
+When the Agent reaches its target, its episode ends and the `OnEpisodeBegin()`
 function moves the target to a random location. In addition, if the Agent rolls
```
Comment (chriselion) [inline docs/Learning-Environment-Create-New.md]:
```suggestion
method moves the target to a random location. In addition, if the Agent rolls
```

File: docs/Learning-Environment-Create-New.md [hunk 6]
```diff
@@ -233,13 +233,13 @@ public override void CollectObservations(VectorSensor sensor)
 }
 ```
 
-The final part of the Agent code is the `Agent.AgentAction()` method, which
+The final part of the Agent code is the `Agent.OnActionReceived()` method, which
 receives the decision from the Brain and assigns the reward.
```
Comment (chriselion) [inline docs/Learning-Environment-Create-New.md]:
```suggestion
receives the actions from the Brain and assigns the reward.
```

File: docs/Learning-Environment-Design-Agents.md [hunk 7]
```diff
@@ -284,7 +284,7 @@ Neither the Policy nor the training algorithm know anything about what the actio
 values themselves mean. The training algorithm simply tries different values for
 the action list and observes the affect on the accumulated rewards over time and
 many training episodes. Thus, the only place actions are defined for an Agent is
-in the `AgentAction()` function. You simply specify the type of vector action
+in the `OnActionReceived()` function. You simply specify the type of vector action
 space, and, for the continuous vector action space, the number of values, and
 then apply the received values appropriately (and consistently) in
 `ActionAct()`.
```
Comment (chriselion) [inline docs/Learning-Environment-Design-Agents.md]: ActionAct?!

File: docs/Learning-Environment-Design.md [hunk 8]
```diff
@@ -40,19 +39,17 @@ use.
...
+2. Calls the `OnEpisodeBegin()` function for each Agent in the scene.
 3. Calls the  `CollectObservations(VectorSensor sensor)` function for each Agent in the scene.
 4. Uses each Agent's Policy to decide on the Agent's next action.
-5. Calls the `AgentAction()` function for each Agent in the scene, passing in
-   the action chosen by the Agent's Policy. (This function is not called if the
-   Agent is done.)
-6. Calls the Agent's `AgentReset()` function if the Agent has reached its `Max
-   Step` count or has otherwise marked itself as `done`.
+5. Calls the `OnActionReceived()` function for each Agent in the scene, passing in
+   the action chosen by the Agent's Policy. (This function is not called if episode is over.)
```
Comment (chriselion) [inline docs/Learning-Environment-Design.md]: "This function is not called if episode is over" - this statement is inaccurate now, right?

**Comment by surfnerd** [comment]
is it possible to leverage [this](https://docs.unity3d.com/Manual/APIUpdater.html). I think it would make our users' lives easier if we could.

File: docs/Getting-Started-with-Balance-Ball.md [hunk 9]
```diff
@@ -60,31 +60,31 @@ behavior:
   determines how an Agent makes decisions. More on Behavior Parameters in
   the next section.
 * **Max Step** — Defines how many simulation steps can occur before the Agent
```
Comment (vincentpierre) [inline docs/Getting-Started-with-Balance-Ball.md]:
```suggestion
* **Max Step** — Defines how many simulation steps can occur before the Agent's
```

File: docs/Getting-Started-with-Balance-Ball.md [hunk 10]
```diff
@@ -60,31 +60,31 @@ behavior:
...
-  decides it is done. In 3D Balance Ball, an Agent restarts after 5000 steps.
+  episode ends. In 3D Balance Ball, an Agent restarts after 5000 steps.
 
 Perhaps the more interesting aspect of an agents is the Agent subclass
 implementation. When you create an Agent, you must extend the base Agent class.
 The Ball3DAgent subclass defines the following methods:
```
Comment (vincentpierre) [inline docs/Getting-Started-with-Balance-Ball.md]:
```suggestion
When you create an Agent, you must extend the base Agent class.
The Ball3DAgent subclass defines the following methods:
```

File: docs/Getting-Started-with-Balance-Ball.md [hunk 11]
```diff
@@ -60,31 +60,31 @@ behavior:
...
+* `Agent.CollectObservations(VectorSensor sensor)` — Called every simulation step. Responsible for
   collecting the Agent's observations of the environment. Since the Behavior
   Parameters of the Agent are set with vector observation
   space with a state size of 8, the `CollectObservations(VectorSensor sensor)` must call
   `VectorSensor.AddObservation()` such that vector size adds up to 8.
-* agent.AgentAction() — Called every simulation step. Receives the action chosen
+* `Agent.OnActionReceived()` — Called every time a new action is ready. Receives the action chosen
   by the Agent. The vector action spaces result in a
-  small change in the agent cube's rotation at each step. The `AgentAction()` function
+  small change in the agent cube's rotation at each step. The `OnActionReceived()` function
```
Comment (vincentpierre) [inline docs/Getting-Started-with-Balance-Ball.md]:
```suggestion
  small change in the agent cube's rotation at each step. The `OnActionReceived()` method
```

File: docs/Getting-Started-with-Balance-Ball.md [hunk 12]
```diff
@@ -60,31 +60,31 @@ behavior:
...
   assigns a reward to the Agent; in this example, an Agent receives a small
   positive reward for each step it keeps the ball on the agent cube's head and a larger,
-  negative reward for dropping the ball. An Agent is also marked as done when it
+  negative reward for dropping the ball. An Agent's episode is also ended when it
   drops the ball so that it will reset with a new ball for the next simulation
   step.
-* agent.Heuristic() - When the `Use Heuristic` checkbox is checked in the Behavior
+* `Agent.Heuristic()` - When the `Use Heuristic` checkbox is checked in the Behavior
```
Comment (vincentpierre) [inline docs/Getting-Started-with-Balance-Ball.md]:
```suggestion
* `Agent.Heuristic()` - When the `Behavior Type` is set to `Heuristic Only` in the Behavior
```

File: com.unity.ml-agents/CHANGELOG.md [hunk 13]
```diff
@@ -16,6 +16,12 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
  - All SideChannel related code has been moved to the namespace `MLAgents.SideChannels`.
  - `BrainParameters` and `SpaceType` have been removed from the public API
  - `BehaviorParameters` have been removed from the public API.
+ - The following methods in the `Agent` class were renamed:
+   - `InitializeAgent()` was renamed to `Initialize()`
+   - `AgentAction()` was renamed to `OnActionReceived()`
+   - `AgentReset()` was renamed to `OnEpsiodeBegin()`
```
Comment (awjuliani) [inline com.unity.ml-agents/CHANGELOG.md]: Typo `OnEpsiodeBegin()` -> `OnEpisodeBegin()`

Comment (vincentpierre) [reply com.unity.ml-agents/CHANGELOG.md]:
```suggestion
   - `AgentReset()` was renamed to `OnEpisodeBegin()`
```

**Comment by vincentpierre** [comment]
> I found a typo, but otherwise looks good.

All right, I will wait for another approval since this is a rather controversial PR.

File: docs/Getting-Started-with-Balance-Ball.md [hunk 14]
```diff
@@ -59,32 +59,31 @@ behavior:
...
   training generalizes to more than a specific starting position and agent cube
   attitude.
-* agent.CollectObservations(VectorSensor sensor) — Called every simulation step. Responsible for
+* `Agent.CollectObservations(VectorSensor sensor)` — Called every simulation step. Responsible for
   collecting the Agent's observations of the environment. Since the Behavior
   Parameters of the Agent are set with vector observation
   space with a state size of 8, the `CollectObservations(VectorSensor sensor)` must call
   `VectorSensor.AddObservation()` such that vector size adds up to 8.
-* agent.AgentAction() — Called every simulation step. Receives the action chosen
+* `Agent.OnActionReceived()` — Called every time a new action is ready. Receives the action chosen
```
Comment (ervteng) [inline docs/Getting-Started-with-Balance-Ball.md]: Could we say "Called every time the Agent receives an action to take" instead? Seems a bit clearer

File: docs/Learning-Environment-Create-New.md [hunk 15]
```diff
@@ -128,8 +128,8 @@ to the cube using reinforcement learning.
 
 ### Initialization and Resetting the Agent
 
-When the Agent reaches its target, it marks itself done and its Agent reset
-function moves the target to a random location. In addition, if the Agent rolls
+When the Agent reaches its target, its episode ends and the `OnEpisodeBegin()`
+method moves the target to a random location. In addition, if the Agent rolls
 off the platform, the reset function puts it back onto the floor.
```
Comment (ervteng) [inline docs/Learning-Environment-Create-New.md]: `the reset function` -> `the OnEpisodeBegin() method`

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 16]
```diff
@@ -665,7 +693,15 @@ public virtual void CollectDiscreteActionMasks(DiscreteActionMasker actionMasker
         /// Vector action. Note that for discrete actions, the provided array
         /// will be of length 1.
         /// </param>
-        public virtual void AgentAction(float[] vectorAction)
+        public virtual void OnActionReceived(float[] vectorAction)
+        {
+#pragma warning disable 0618
+                AgentAction(m_Action.vectorActions);
```
Comment (mmattar) [inline com.unity.ml-agents/Runtime/Agent.cs]: indentation is off here.

File: docs/Migrating.md [hunk 17]
```diff
@@ -24,6 +24,12 @@ The versions can be found in
 * `BrainParameters` and `SpaceType` have been removed from the public API
 * `BehaviorParameters` have been removed from the public API.
 * `DecisionRequester` has been made internal (you can still use the DecisionRequesterComponent from the inspector). `RepeatAction` was renamed `TakeActionsBetweenDecisions` for clarity.
+* The following methods in the `Agent` class were renamed:
```
Comment (mmattar) [inline docs/Migrating.md]: Should we clarify that they've been deprecated and will be removed in a later version?

File: docs/Migrating.md [hunk 18]
```diff
@@ -33,7 +39,13 @@ The versions can be found in
 * If you call `RayPerceptionSensor.PerceiveStatic()` manually, add your inputs to a `RayPerceptionInput`. To get the previous float array output,
  iterate through `RayPerceptionOutput.rayOutputs` and call `RayPerceptionOutput.RayOutput.ToFloatArray()`.
 * Re-import all of your `*.NN` files to work with the updated Barracuda package.
-* Replace all calls to `Agent.GetStepCount()` with `Agent.StepCount`.
+* Replace all calls to `Agent.GetStepCount()` with `Agent.StepCount`
+* Replace the following method with their new equivalent:
```
Comment (mmattar) [inline docs/Migrating.md]: Should we clarify that this is a strong recommendation as these methods will be removed in a later release.

File: com.unity.ml-agents/CHANGELOG.md [hunk 19]
```diff
@@ -16,6 +16,12 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
  - All SideChannel related code has been moved to the namespace `MLAgents.SideChannels`.
  - `BrainParameters` and `SpaceType` have been removed from the public API
  - `BehaviorParameters` have been removed from the public API.
+ - The following methods in the `Agent` class were renamed:
```
Comment (mmattar) [inline com.unity.ml-agents/CHANGELOG.md]: Should we clarify that they've been deprecated and will be removed in a later version?

File: docs/Getting-Started-with-Balance-Ball.md (hunk 14 above)
Comment (vincentpierre) [reply docs/Getting-Started-with-Balance-Ball.md]:
```suggestion
* `Agent.OnActionReceived()` — Called every time the Agent receives an action to take. Receives the action chosen
```

File: docs/Learning-Environment-Create-New.md (hunk 15 above)
Comment (vincentpierre) [inline docs/Learning-Environment-Create-New.md]:
```suggestion
off the platform, the `OnEpisodeBegin()` method puts it back onto the floor.
```

File: com.unity.ml-agents/Runtime/Agent.cs (hunk 16 above)
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Agent.cs]:
```suggestion
        AgentAction(m_Action.vectorActions);
```

File: docs/Migrating.md [hunk 20]
```diff
@@ -33,7 +39,13 @@ The versions can be found in
 * If you call `RayPerceptionSensor.PerceiveStatic()` manually, add your inputs to a `RayPerceptionInput`. To get the previous float array output,
  iterate through `RayPerceptionOutput.rayOutputs` and call `RayPerceptionOutput.RayOutput.ToFloatArray()`.
 * Re-import all of your `*.NN` files to work with the updated Barracuda package.
-* Replace all calls to `Agent.GetStepCount()` with `Agent.StepCount`.
+* Replace all calls to `Agent.GetStepCount()` with `Agent.StepCount`
+* We strongly recommend replacing the following method with their new equivalent as they will be removed in a later release:
```
Comment (ervteng) [inline docs/Migrating.md]: `following methods`

Comment (vincentpierre) [reply docs/Migrating.md]:
```suggestion
* We strongly recommend replacing the following methods with their new equivalent as they will be removed in a later release:
```

File: docs/Migrating.md [hunk 21]
```diff
@@ -24,6 +24,12 @@ The versions can be found in
 * `BrainParameters` and `SpaceType` have been removed from the public API
 * `BehaviorParameters` have been removed from the public API.
 * `DecisionRequester` has been made internal (you can still use the DecisionRequesterComponent from the inspector). `RepeatAction` was renamed `TakeActionsBetweenDecisions` for clarity.
+* The following methods in the `Agent` class have been deprecated. They will be removed in a later release:
```
Comment (ervteng) [inline docs/Migrating.md]: Can we change this to something like "The following methods in the `Agent` class have been renamed. The original method names will be removed in a later release."

The current sentence sounds like all the methods will be deprecated (including the renamed one)

Comment (vincentpierre) [reply docs/Migrating.md]:
```suggestion
* The following methods in the `Agent` class have been renamed. The original method names will be removed in a later release:
```

**Pull request merged**